src/peft/tuners/vera/model.py

```diff
@@ -0,0 +1,564 @@
+        >>> model = AutoModelForSeq2SeqLM.from_pretrained("t5-base")
+        >>> vera_model = VeraModel(model, config, "default")
+        ```
+
+        ```py
+        >>> import transformers
+        >>> from peft import VeraConfig, PeftModel, get_peft_model
+
+        >>> target_modules = ["q_proj", "k_proj", "v_proj", "out_proj", "fc_in", "fc_out", "wte"]
+        >>> config = VeraConfig(
+        ...     r=4, target_modules=target_modules, vera_dropout=0.1, bias="none", task_type="CAUSAL_LM"
+        ... )
+
+        >>> model = transformers.GPTJForCausalLM.from_pretrained(
+        ...     "kakaobrain/kogpt",
+        ...     revision="KoGPT6B-ryan1.5b-float16",  # or float32 version: revision=KoGPT6B-ryan1.5b
+        ...     pad_token_id=tokenizer.eos_token_id,
+        ...     use_cache=False,
+        ...     device_map={"": rank},
+        ...     torch_dtype=torch.float16,
+        ...     load_in_8bit=True,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> vera_model = get_peft_model(model, config)
+        ```
+
+    **Attributes**:
+        - **model** ([`~transformers.PreTrainedModel`]) -- The model to be adapted.
+        - **peft_config** ([`VeraConfig`]): The configuration of the Vera model.
+    """
+
+    def _find_first_dim(self) -> int:
+        """
+        Finds the first linear and embedding that has been wrapped with Vera, and extract the input and output
+        dimension.
+
+        This will be used for determining the size of the shared vera_A and vera_B matrices.
+
+        This will throw an error if there are multiple layers of the same type with different shapes.
+        """
+        first_linear, first_embedding = None, None
+        for module in self.model.modules():
+            if isinstance(module, Linear):
+                module_shape = tuple(module.weight.shape)
+                if module.fan_in_fan_out:
+                    module_shape = module_shape[::-1]
+
+                if first_linear is not None and module_shape != first_linear:
+                    raise ValueError(
+                        "Multiple target linear layers with different dimensions were specified! Vera only supports a"
+                        f" single dimension size. Got '{module_shape}' expected '{first_linear}"
+                    )
+                first_linear = module_shape
+
+            elif isinstance(module, Embedding):
+                if first_embedding is not None and tuple(module.weight.shape) != first_embedding:
+                    raise ValueError(
+                        "Multiple target embedding layers with different dimensions or vocabulary sizes were"
+                        " specified! Vera only supports a single size."
+                    )
+                first_embedding = tuple(module.weight.shape)
+
+        return first_linear, first_embedding
+
+    def __init__(self, model, config, adapter_name) -> None:
+        if config.projection_prng_key is None:
```

Comment:
I think it's better to run this check directly inside of the config's `__post_init__` to give feedback as early as possible. The docstring for `projection_prng_key` should also be adjusted to say it cannot be `None`, even if that's the default value.